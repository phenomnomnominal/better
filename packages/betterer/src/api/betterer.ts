import type { BettererSuiteSummary } from '../suite/index.js';
import type { BettererOptions } from './types.js';

import { BettererRunnerΩ } from '../runner/index.js';
import { merge } from './merge.js';
import { results } from './results.js';
import { runner } from './runner.js';
import { watch } from './watch.js';

/**
 * @public Run **Betterer** with the given options.
 *
 * @example
 * ```typescript
 * import { betterer } from '@betterer/betterer';
 *
 * const suiteSummary = await betterer(options);
 * ```
 *
 * @param options - Options for running **Betterer**.
 * @throws {@link @betterer/errors#BettererError | `BettererError` }
 * Will throw if something goes wrong while running **Betterer**.
 */
export const betterer = async function betterer(options: BettererOptions = {}): Promise<BettererSuiteSummary> {
  const runner = await BettererRunnerΩ.create(options);
  return await runner.run();
};

betterer.merge = merge;
betterer.results = results;
betterer.runner = runner;
betterer.watch = watch;
