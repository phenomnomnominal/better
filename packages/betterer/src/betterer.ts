import { debug } from '@phenomnomnominal/debug';

import { BettererBaseConfigPartial, BettererStartConfigPartial, BettererWatchConfigPartial } from './config';
import { BettererSummary } from './context';
import { BettererRunner, BettererRunnerΩ, BettererWatcherΩ } from './runner';

export async function betterer(partialConfig: BettererStartConfigPartial = {}): Promise<BettererSummary> {
  initDebug();
  const runner = new BettererRunnerΩ();
  await runner.start(partialConfig);
  await runner.queue();
  return runner.stop();
}

export async function runner(partialConfig?: BettererBaseConfigPartial): Promise<BettererRunner> {
  initDebug();
  const runner = new BettererRunnerΩ();
  await runner.start(partialConfig);
  return runner;
}
betterer.runner = runner;

export async function watch(partialConfig: BettererWatchConfigPartial = {}): Promise<BettererRunner> {
  initDebug();
  partialConfig.watch = true;
  const runner = new BettererRunnerΩ();
  const context = await runner.start(partialConfig);
  const watcher = new BettererWatcherΩ(context, runner);
  await watcher.setup();
  return watcher;
}
betterer.watch = watch;

function initDebug(): void {
  const enabled = !!process.env.BETTERER_DEBUG;
  if (enabled) {
    debug({
      header: 'betterer',
      include: [/@betterer\//],
      ignore: [require.resolve('./utils')],
      enabled,
      time: !!process.env.BETTERER_DEBUG_TIME,
      values: !!process.env.BETTERER_DEBUG_VALUES,
      logPath: process.env.BETTERER_DEBUG_LOG
    });
  }
}
