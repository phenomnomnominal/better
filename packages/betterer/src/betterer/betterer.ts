import logDiff from 'jest-diff';

import { BettererGoal, BettererGoalFunction, BettererTest, BettererConstraint, BettererDiff } from './types';

export type BettererOptions<TestType, SerialisedType = TestType> = {
  test: BettererTest<TestType>;
  constraint: BettererConstraint<SerialisedType>;
  goal: BettererGoal<SerialisedType>;
  diff: BettererDiff<TestType, SerialisedType>;
  isOnly?: boolean;
  isSkipped?: boolean;
};

export class Betterer<Base = unknown, Serialised = Base> {
  public constraint: BettererConstraint<Serialised>;
  public diff: BettererDiff<Base, Serialised>;
  public goal: BettererGoalFunction<Serialised>;
  public test: BettererTest<Base>;
  public isOnly = false;
  public isSkipped = false;
  public name = '';

  constructor({
    test,
    constraint,
    goal,
    diff = defaultDiff,
    isOnly = false,
    isSkipped = false
  }: BettererOptions<Base, Serialised>) {
    this.constraint = constraint;
    this.diff = diff;
    this.goal = createGoal(goal);
    this.test = test;
    this.isOnly = isOnly;
    this.isSkipped = isSkipped;
  }

  public only(): Betterer<Base, Serialised> {
    this.isOnly = true;
    return this;
  }

  public skip(): Betterer<Base, Serialised> {
    this.isSkipped = true;
    return this;
  }
}

function createGoal(goal: BettererGoal<unknown>): BettererGoalFunction<unknown> {
  if (typeof goal === 'function') {
    return goal as BettererGoalFunction<unknown>;
  }
  return (value: unknown): boolean => value === goal;
}

function defaultDiff(_: unknown, serialisedCurrent: unknown, serialisedPrevious: unknown): void {
  const diffStr =
    logDiff(serialisedPrevious, serialisedCurrent, {
      aAnnotation: 'Previous',
      bAnnotation: 'Current'
    }) || '';
  console.log(
    diffStr
      .split('\n')
      .map((line) => `  ${line}`)
      .join('\n')
  );
}
