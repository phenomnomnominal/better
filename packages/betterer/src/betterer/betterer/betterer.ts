import { BettererRun } from '../../context';
import { BettererConstraint, BettererGoal, BettererTestFunction, BettererOptions } from './types';

export class Betterer<TestType = unknown> {
  public readonly constraint: BettererConstraint<TestType>;
  public readonly goal: BettererGoal<TestType>;
  public readonly test: BettererTestFunction<TestType>;

  public readonly isBetterer = true;

  private _isOnly: boolean;
  private _isSkipped: boolean;

  constructor(options: BettererOptions<TestType>) {
    const { constraint, test, isOnly, isSkipped } = options;
    this.constraint = constraint;
    this.goal = this._createGoal(options);
    this.test = test;
    this._isOnly = isOnly || false;
    this._isSkipped = isSkipped || false;
  }

  public get isOnly(): boolean {
    return this._isOnly;
  }

  public get isSkipped(): boolean {
    return this._isSkipped;
  }

  public only(): this {
    this._isOnly = true;
    return this;
  }

  public skip(): this {
    this._isSkipped = true;
    return this;
  }

  public getExpected(run: BettererRun): TestType {
    return run.test.expected as TestType;
  }

  private _createGoal(options: BettererOptions<TestType>): BettererGoal<TestType> {
    const hasGoal = Object.hasOwnProperty.call(options, 'goal');
    if (!hasGoal) {
      return (): boolean => false;
    }
    const { goal } = options;
    if (this._isGoalFunction(goal)) {
      return goal;
    }
    return (value: unknown): boolean => value === goal;
  }

  private _isGoalFunction(goal: unknown): goal is BettererGoal<unknown> {
    return typeof goal === 'function';
  }
}

export function isBetterer(obj: unknown): obj is Betterer {
  return !!(obj as Betterer).isBetterer;
}
