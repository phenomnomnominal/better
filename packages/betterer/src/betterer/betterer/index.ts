export * from './betterer';
export * from './types';
