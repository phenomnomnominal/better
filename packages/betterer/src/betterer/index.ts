export { createBetterer } from './create-betterer';
export { Betterer, isBetterer } from './betterer';
export { BettererFiles, BettererFile, FileBetterer, createFileBetterer, isFileBetterer } from './file-betterer';
export { SerialisableBetterer, isSerialisableBetterer } from './serialisable-betterer';
export * from './betterer/types';
export * from './file-betterer/types';
export * from './serialisable-betterer/types';
export * from './types';
