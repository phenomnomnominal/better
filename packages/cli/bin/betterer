#!/usr/bin/env node
import { cli__ } from '../dist/cli.js';
cli__(process.cwd(), process.argv);
