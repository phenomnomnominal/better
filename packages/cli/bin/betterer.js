#!/usr/bin/env node
import { cliΔ } from '../dist/cli.js';
void cliΔ(globalThis.process.cwd(), globalThis.process.argv);
