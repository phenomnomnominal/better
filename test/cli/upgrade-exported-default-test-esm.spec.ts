// eslint-disable-next-line require-extensions/require-extensions -- tests not ESM ready yet
import { createFixture } from '../fixture';

const ARGV = ['node', './bin/betterer', 'upgrade'];

describe('betterer upgrade', () => {
  it('should upgrade exported default test in an ES module', async () => {
    const { cleanup, logs, paths, readFile } = await createFixture(
      'upgrade-exported-default-test-esm',
      {
        './.betterer.ts': `
import { BettererTest, BettererFileTest } from '@betterer/betterer';
import { bigger } from '@betterer/constraints';

let start = 0;

export default {
  getsBetter: {
    test: () => start++,
    constraint: bigger
  },
  'gets better': new BettererTest({
    test: () => start++,
    constraint: bigger
  }),
  countFiles: new BettererFileTest(async (files, fileTestResult) => {        
    const [filePath] = files;
    const file = fileTestResult.addFile(filePath, '');
    file.addIssue(0, 0, '\`$' + '{key}\`');
  }),
  'count files': new BettererFileTest(async (files, fileTestResult) => {        
    const [filePath] = files;
    const file = fileTestResult.addFile(filePath, '');
    file.addIssue(0, 0, '\`$' + '{key}\`');
  }).include('./src/**/*.ts')
}
        `
      },
      {
        logFilters: [/ðŸŒŸ Upgrading Betterer/]
      }
    );

    const fixturePath = paths.cwd;

    const { cli__ } = await import('@betterer/cli');

    await cli__(fixturePath, ARGV);

    await cli__(fixturePath, [...ARGV, '--save']);

    const upgradedConfig = await readFile(`${paths.config}.ts`);

    expect(upgradedConfig).toMatchSnapshot();

    expect(logs).toMatchSnapshot();

    await cleanup();
  });
});
