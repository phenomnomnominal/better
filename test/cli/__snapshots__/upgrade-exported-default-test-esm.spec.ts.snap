// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`betterer upgrade > should upgrade exported default test in an ES module 1`] = `
"import { BettererTest, BettererFileTest } from "@betterer/betterer";
import { bigger } from "@betterer/constraints";

let start = 0;

export default {
  getsBetter: () =>
    new BettererTest({
      test: () => start++,
      constraint: bigger,
    }),
  "gets better": () =>
    new BettererTest({
      test: () => start++,
      constraint: bigger,
    }),
  countFiles: () =>
    new BettererFileTest(async (files, fileTestResult) => {
      const [filePath] = files;
      const file = fileTestResult.addFile(filePath, "");
      file.addIssue(0, 0, "\`$" + "{key}\`");
    }),
  "count files": () =>
    new BettererFileTest(async (files, fileTestResult) => {
      const [filePath] = files;
      const file = fileTestResult.addFile(filePath, "");
      file.addIssue(0, 0, "\`$" + "{key}\`");
    }).include("./src/**/*.ts"),
};
"
`;

exports[`betterer upgrade > should upgrade exported default test in an ES module 2`] = `
[
  "🎉 Upgrading Betterer (0ms): 1 task done!
✅ Upgrading "./.betterer.ts": done!
・
・ Betterer can upgrade your test definition file automatically! ✨
・ Here's what the changes will look like:
・
・ - Before
・ + After
・
・ - import { BettererTest, BettererFileTest } from '@betterer/betterer';
・ + import { BettererTest, BettererFileTest } from "@betterer/betterer";
・ - import { bigger } from '@betterer/constraints';
・ + import { bigger } from "@betterer/constraints";
・
・   let start = 0;
・
・   export default {
・ -   getsBetter: {
・ +   getsBetter: () =>
・ +     new BettererTest({
・ -     test: () => start++,
・ +       test: () => start++,
・ -     constraint: bigger
・ -   },
・ +       constraint: bigger,
・ +     }),
・ -   'gets better': new BettererTest({
・ +   "gets better": () =>
・ +     new BettererTest({
・ -     test: () => start++,
・ +       test: () => start++,
・ -     constraint: bigger
・ -   }),
・ +       constraint: bigger,
・ +     }),
・ -   countFiles: new BettererFileTest(async (files, fileTestResult) => {
・ -     const [filePath] = files;
・ +   countFiles: () =>
・ +     new BettererFileTest(async (files, fileTestResult) => {
・ +       const [filePath] = files;
・ -     const file = fileTestResult.addFile(filePath, '');
・ +       const file = fileTestResult.addFile(filePath, "");
・ -     file.addIssue(0, 0, '\`$' + '{key}\`');
・ +       file.addIssue(0, 0, "\`$" + "{key}\`");
・ -   }),
・ +     }),
・ -   'count files': new BettererFileTest(async (files, fileTestResult) => {
・ -     const [filePath] = files;
・ +   "count files": () =>
・ +     new BettererFileTest(async (files, fileTestResult) => {
・ +       const [filePath] = files;
・ -     const file = fileTestResult.addFile(filePath, '');
・ +       const file = fileTestResult.addFile(filePath, "");
・ -     file.addIssue(0, 0, '\`$' + '{key}\`');
・ +       file.addIssue(0, 0, "\`$" + "{key}\`");
・ -   }).include('./src/**/*.ts')
・ - }
・ +     }).include("./src/**/*.ts"),
・ + };
・ +
・
・ If that looks okay, run \`betterer upgrade --save\` to write the changes to disk. 💾
・",
  "🎉 Upgrading Betterer (0ms): 1 task done!
✅ Upgrading "./.betterer.ts": done!
・ Saved upgraded config to "fixtures/upgrade-exported-default-test-esm/.betterer.ts"! ☀️",
]
`;
